### PR TITLE
Full website update

### PR DESCRIPTION
**Pull request by Turbotailz**
Bunch of work on the full LuckPerms website upgrade, which will hopefully replace the existing site at `luckperms.net`.

This update brings in the following:

- Home page
- Downloads page (with quiz)
- Verbose viewer
- Tree viewer
- Wiki section (fully functional and uses LuckPerms/wiki as a data source

A lot of refactoring and styling has gone on behind the scenes. Everything is now contained with a max-width so wide-screen users have a better experience. Mobile responsiveness will come later.

Ideally this replaces the existing web services, measures have been taken to ensure older versions of LuckPerms will still work with at least the verbose/tree viewer. The web editor might not still work for version 4 users, may need to figure out a solution there. For v5 users, maybe make `editor.luckperms.net/:id` redirect to `luckperms.net/editor/:id`?

Special thanks to @Tobi406 for the PR with the wiki addition 👌 

Live version of this code is available at https://luckperms.turbotailz.com

### CODE REVIEW

File: src/views/Wiki.vue [hunk 1]
```diff
@@ -0,0 +1,79 @@
+<template>
+  <main class="wiki container">
+    <aside>
+      <h1>Wiki</h1>
+      <sidebar />
```
Comment (Tobi406) [inline src/views/Wiki.vue:5]:
```suggestion
      <sidebar />
      <br />
```

The last link is already pretty far below.
grafik

Maybe consider adding a `<br />`? Looks like this then:
grafik

Comment (Turbotailz) [reply src/views/Wiki.vue:5]: Hmm doesn't display like that for me. Is that Firefox?

Comment (Laarryy) [reply src/views/Wiki.vue:5]: Displays like that for me too on Firefox!

Comment (Turbotailz) [reply src/views/Wiki.vue:5]: Classic Firefox, at least it's not Safari 😝

**Comment by Tobi406** [comment]
Just another idea: As there is already a demo for the editor (https://editor.luckperms.net/demo), why don't link it from the editor subpage, as that is now linked from the navbar?
And what about demos for the tree and verbose viewer.

**Comment by Turbotailz** [comment]
Yeah I was going to add some demo data to them.

File: src/views/Download.vue [hunk 2]
```diff
@@ -1,77 +1,97 @@
...
+          <ol>
+            <li>Place the downloaded <code>.jar</code> file into your server's <code>plugins</code> or <code>mods</code> folder</li>
+            <li>Start or restart your server - <strong>do not reload!</strong></li>
+            <li>Check your <code>plugins</code> or <code>mods</code> folder for a <code>LuckPerms</code> folder - see the tips below if the folder doesn't generate</li>
+            <li>Open the <code>config</code> file in your favourite text editor and adjust any settings to your liking</li>
+          </ol>
+          <h2>Having trouble installing?</h2>
+          <ul>
+            <li>Make sure to check your console for any errors - especially during start up</li>
+            <li>Check the <a href="https://github.com/lucko/LuckPerms/wiki" target="_blank">wiki</a> to see if you missed any important steps during setup</li>
```
Comment (Tobi406) [inline src/views/Download.vue:77]: Link to `/wiki`?

Comment (Turbotailz) [reply src/views/Download.vue:77]: `<router-link to="/wiki"> ... </router-link>`

File: src/views/Home.vue [hunk 3]
```diff
@@ -1,106 +1,111 @@
...
+              Verbose Viewer
+            </router-link>
+            <router-link to="/tree" alt="Tree Viewer">
+              <font-awesome icon="sitemap" />
+              Tree Viewer
+            </router-link>
+          </div>
+        </div>
+        <div>
+          <a href="/wiki" class="resource">
```
Comment (Tobi406) [inline src/views/Home.vue:56]: Maybe `<routerlink>` here too?

File: vue.config.js [hunk 4]
```diff
@@ -3,9 +3,32 @@ module.exports = {
...
+      .loader('vue-markdown-loader/lib/markdown-compiler')
+      .options({
+        raw: true,
+        linkify: true,
+        use: [
+          [require('markdown-it-anchor'), {
+            "permalink": true,
+            "permalinkSymbol": "🔗"
+          }],
+          require('markdown-it-anchor'),
```
Comment (Tobi406) [inline vue.config.js:29]: Duplicated markdown-it-anchor here

File: src/views/Wiki.vue [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+      margin-bottom: 2rem;
+    }
+
+    h1 {
+      font-size: 2rem;
+      margin-top: 0;
+    }
+
+    a {
+      text-decoration: none;
```
Comment (Tobi406) [inline src/views/Wiki.vue:64]:
```suggestion
      text-decoration: none;
      margin: .2rem;
```

To have the outlines show better :)

File: src/views/Download.vue [hunk 6]
```diff
@@ -1,77 +1,130 @@
...
+            <small>LuckPerms 5.0 and above</small>
+          </a>
+				</div>
+				
+				<div>
+					<p>Allows for other ways to make <router-link to="/wiki/Default-Groups">Default Groups</router-link> if the workarounds are not possible.</p>
+          <p>Check out the <router-link to="/wiki/Extensions#extension-default-assignments">wiki section</router-link> for more information! See also <a href="/wiki/Default-Groups#configure-default-assignments">this section</a> about configuring default assignments!</p>
+				</div>
+			</section>
+    </div>
```
Comment (Tobi406) [inline src/views/Download.vue:127]: Fix tabs around here.

**Comment by lucko** [comment]
Overall, so so good. A massive improvement to what we had before :tada: :tada:

Just a few suggestions/ideas from me though:

Generally, I feel like all content on the site could do with being 80% of the size - especially when using the editor. For example, [zoom at 100%](https://i.imgur.com/KeKtGJf.png) vs [zoom at 80%](https://i.imgur.com/XLZzpe7.png) - hopefully you can see what I mean!

The OpenGraph and Twitter card meta tags are currently missing - not sure if that's intentional - but would be cool if they could be re-added - varied for each page ofc but here's the ones for the old homepage:
https://github.com/lucko/LuckPermsWeb/blob/c6911a3db7d28fa1b0a47c9c9d53116595cf8146/index.html#L11-L21

Also:

* Remove the version string from the top left of the homepage - that's used to version the plugin jar, not the website!
* Download page
  * Would be better imo if the download buttons were all on screen immediately without having to scroll - currently only the Bukkit one is visible on my display with the browser full screen ([img](https://i.imgur.com/IssrgNr.png)) - perhaps the individual download buttons could be spread out horizontally? 
  * The "`Download LuckPerms` `Current version: 5.0.139` `Not sure which version?`" bit could probably be a bit smaller - the emphasis on this page should be on the links to download a jar.
  * In the expansions section, it's not obvious that the descriptions on the right relate to the downloads on the right - again maybe listing them horizontally with the descriptions underneath could be better.
* Wiki
  * Would be awesome if more of the sidebar could be visible (on screen) without scrolling - less padding around the title & headers perhaps?

File: src/App.vue [hunk 7]
```diff
@@ -1,9 +1,78 @@
...
-      <router-link to="/editor">Editor</router-link>
-    </div> -->
+    <div id="nav">
+      <div class="container">
+        <div>
+          <router-link to="/" class="logo">
+            <img alt="LuckPerms logo" src="@/assets/logo.png">
+            <span>LuckPerms</span>
+          </router-link>
+          <span v-if="version">v{{ version }}</span>
```
Comment (Tobi406) [inline src/App.vue:10]:
```suggestion
```

Remove version at the top.

**Comment by Tobi406** [comment]
About the version thing, we can make it into two rows, then it's visible (atleast for me) without scrolling. And we can still reduce some space here and there.

image

**Comment by Tobi406** [comment]
What about the missing links we haven't added?
- The Jenkins one
- The resource pages

I mean these: https://imgur.com/a/KoezQWf

**Comment by Tobi406** [comment]
And regarding the extensions, we could maybe just use some more space between them? https://imgur.com/a/KBr5KLl

**Comment by Andre601** [comment]
One thing I was wondering about if it is possible...
But maybe have the demo (https://editor.luckperms.net/demo) be a bit more interactive?
Like the user could click a button to start a guide on how the editor works by having him do stuff like creating a new group, deleting the old one, setting permissions with context, etc.
That way would the user learn all the basics of the editor like how to set a world context.

**Comment by Turbotailz** [comment]
It seems that most of the issues you raised @lucko are just display responsiveness which is something I have been meaning to work on, so I'll make that my next priority to make the site/tools display better on smaller screens 👍 

The meta tags will definitely be added.

The version number was never meant to indicate the website version, though I can see how it can be understood that way. I'll commit Tobi's change above.

> But maybe have the demo be a bit more interactive?

I really like this idea and *have* thought about it, but it just seems like it could be a lot of work. Will put it on the backlog.

File: src/views/Tree.vue [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+            <a href="/tree/demo"><button class="button demo-button">View Demo</button></a>
+            <p>To generate a permission tree, do the following in game or from the console:</p>
+            <ul>
+              <li><code>/lp tree [scope] [player]</code></li>
+              <li>Follow the URL that is generated</li>
+            </ul>
+          </template>
+          <div v-else class="error">
+            <p><strong>There was an error loading the data.</strong> Either the URL was copied wrong or the session has expired.</p>
+            <p>Please generate another editor session with <code>/lp editor</code>.</p>
```
Comment (Tobi406) [inline src/views/Tree.vue:59]:
```suggestion
            <p>Please generate another editor session with <code>/lp tree</code>.</p>
```

File: src/views/Verbose.vue [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+            <ul>
+              <li><code>/lp verbose record [filter]</code></li>
+              <li>Perform a series of actions that require permissions</li>
+              <li><code>/lp verbose paste</code></li>
+              <li>Follow the URL that is generated</li>
+            </ul>
+          </template>
+          <div v-if="errors.load" class="error">
+            <p><strong>There was an error loading the data.</strong> Either the URL was copied wrong or the session has expired.</p>
+            <p>Please generate another editor session with <code>/lp editor</code>.</p>
```
Comment (Tobi406) [inline src/views/Verbose.vue:97]:
```suggestion
            <p>Please generate another editor session with <code>/lp verbose</code>.</p>
```

File: src/App.vue [hunk 10]
```diff
@@ -35,6 +137,39 @@
...
+  data() {
+    return {
+      menu: false,
+    };
+  },
+
+  computed: {
+    version() {
+      return this.$store.getters.version;
+    },
```
Comment (Tobi406) [inline src/App.vue:151]: Is this even needed anymore?

Comment (Turbotailz) [reply src/App.vue:151]: Ah yeah, that was to show the version in the navigation which has been removed. So I guess this can be removed, also.

**Comment by Turbotailz** [comment]
Main things to do before this can be merged into master:

- [ ] handle redirect from v4 sessions with `#` in the URL
- [ ] add meta data
- [ ] fix various responsive issues (not essential for going live but nice to have)
- [ ] add top secret page 👀 
- [ ] verbose/tree pages may need optimising for larger sets of data (need some test data)

Maybe other things that I've missed, will update if so.

**Comment by Turbotailz** [comment]
Verbose viewer needs some serious improving for really large datasets (like over 100,000 checks) but I'm having trouble with finding a solution. Will mark this as ready for merging now.

**Comment by Tobi406** [comment]
There are still some responsive issues for mobile devices I think: 
- The hero background doesn't cover everything. Not a mayor issue tough. image

- The wiki sidebar covers everything (and more) and no article is displayed at all.
image

File: .gitmodules [hunk 11]
```diff
@@ -0,0 +1,3 @@
+[submodule "src/wiki"]
+	path = src/wiki
+	url = https://github.com/Turbotailz/wiki
```
Comment (Tobi406) [inline .gitmodules:3]: Dont't forget to change the url here after LuckPerms/wiki#57 was merged.

**Pull request merged**